pkg/resources/gce/gce.go

```diff
@@ -498,89 +498,153 @@ func (d *clusterDiscoveryGCE) listFirewallRules() ([]*resources.Resource, error)
+		if len(firewallRule.TargetTags) == 0 {
+			continue nextFirewallRule
+		}
+
+		firewallRuleResource := &resources.Resource{
+			Name:    firewallRule.Name,
+			ID:      firewallRule.Name,
+			Type:    typeFirewallRule,
+			Deleter: deleteFirewallRule,
+			Obj:     firewallRule,
+		}
+		firewallRuleResource.Blocks = append(firewallRuleResource.Blocks, typeNetwork+":"+gce.LastComponent(firewallRule.Network))
+
+		if d.matchesClusterNameMultipart(firewallRule.Name, maxPrefixTokens) {
+			klog.V(4).Infof("Found resource: %s", firewallRule.SelfLink)
+			resourceTrackers = append(resourceTrackers, firewallRuleResource)
 		}
 
-		// find the Kubernetes LoadBalancer
-		if strings.HasPrefix(fr.Name, "k8s-fw-") {
-			name := strings.ReplaceAll(fr.Name, "k8s-fw-", "")
-			fr, err := c.Compute().ForwardingRules().Get(c.Project(), c.Region(), name)
+		// find the objects if this is a Kubernetes LoadBalancer
+		if strings.HasPrefix(firewallRule.Name, "k8s-fw-") {
+			// We build a list of resources if this is a k8s firewall rule,
+			// but we only add them once all the checks are complete
+			var k8sResources []*resources.Resource
+
+			k8sResources = append(k8sResources, firewallRuleResource)
+
+			// We lookup the forwarding rule by name, but we then validate that it points to one of our resources
+			forwardingRuleName := strings.TrimPrefix(firewallRule.Name, "k8s-fw-")
+			forwardingRule, err := c.Compute().ForwardingRules().Get(c.Project(), c.Region(), forwardingRuleName)
 			if err != nil {
-				return nil, fmt.Errorf("error get ForwardingRule: %v", err)
+				if gce.IsNotFound(err) {
+					// We looked it up by name, so an error isn't unlikely
+					klog.Warningf("could not find forwarding rule %q, assuming firewallRule %q is not a k8s rule", forwardingRuleName, firewallRule.Name)
+					continue nextFirewallRule
+				}
+				return nil, fmt.Errorf("error getting ForwardingRule %q: %w", forwardingRuleName, err)
 			}
 
-			frResourceTracker := &resources.Resource{
-				Name:    fr.Name,
-				ID:      fr.Name,
+			forwardingRuleResource := &resources.Resource{
+				Name:    forwardingRule.Name,
+				ID:      forwardingRule.Name,
 				Type:    typeForwardingRule,
 				Deleter: deleteForwardingRule,
-				Obj:     fr,
+				Obj:     forwardingRule,
 			}
-			if fr.Target != "" {
-				frResourceTracker.Blocks = append(frResourceTracker.Blocks, typeTargetPool+":"+gce.LastComponent(fr.Target))
+			if forwardingRule.Target != "" {
+				forwardingRuleResource.Blocks = append(forwardingRuleResource.Blocks, typeTargetPool+":"+gce.LastComponent(forwardingRule.Target))
 			}
-			resourceTrackers = append(resourceTrackers, frResourceTracker)
+			k8sResources = append(k8sResources, forwardingRuleResource)
 
-			tp, err := c.Compute().TargetPools().Get(c.Project(), c.Region(), name)
+			// TODO: Can we get k8s to set labels on the ForwardingRule?
+
+			// TODO: Check description?  It looks like e.g. description: '{"kubernetes.io/service-name":"kube-system/guestbook"}'
```

Comment:
yep but what if we have multiple clusters in same project? GCP cloudprovider should have something similar like CSI has "custom labels" https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver#createvolume-parameters

However, these things are not going to happen before we have 1) gce external ccm in use 2) after that custom labels should be implemented